checkov/terraform/tf_parser.py

```diff
@@ -329,6 +330,67 @@ def parse_hcl_module(
 
         return module, tf_definitions
 
+    '''
+    This function is similar to parse_hcl_module, except that it creates a list of tuples instead of a single tuple.
+    The objective is to create a collection of TF definitions based on directory, instead of a single big structure.
+    This will allow us to boost performance by running on several smaller objects rather than one.
+    '''
+    def parse_multi_graph_hcl_module(
+        self,
+        source_dir: str,
+        source: str,
+        download_external_modules: bool = False,
+        external_modules_download_path: str = DEFAULT_EXTERNAL_MODULES_DIR,
+        parsing_errors: dict[str, Exception] | None = None,
+        excluded_paths: list[str] | None = None,
+        vars_files: list[str] | None = None,
+        external_modules_content_cache: dict[str, ModuleContent | None] | None = None,
+        create_graph: bool = True,
+    ) -> list[tuple[Module, list[dict[TFDefinitionKey, dict[str, Any]]]]]:
+        tf_definitions = self.parse_directory(
+            directory=source_dir, out_evaluations_context={},
+            out_parsing_errors=parsing_errors if parsing_errors is not None else {},
+            download_external_modules=download_external_modules,
+            external_modules_download_path=external_modules_download_path, excluded_paths=excluded_paths,
+            vars_files=vars_files, external_modules_content_cache=external_modules_content_cache
+        )
+        tf_definitions = clean_parser_types(tf_definitions)
+        tf_definitions = serialize_definitions(tf_definitions)
+
+        dirs_to_definitions = self.create_definition_by_dirs(tf_definitions)
+
+        modules_and_definitions_tuple: list[tuple[Module, list[dict[TFDefinitionKey, dict[str, Any]]]]] = []
+        if create_graph:
+            for source_path, definitions in dirs_to_definitions.items():
+                module, parsed_tf_definitions = self.parse_hcl_module_from_multi_tf_definitions(definitions, source_path, source)
+                modules_and_definitions_tuple.append((module, parsed_tf_definitions))
+
+        return modules_and_definitions_tuple
+
+    def create_definition_by_dirs(self, tf_definitions: dict[TFDefinitionKey, dict[str, list[dict[str, Any]]]]
```

Comment:
I would add specific unit tests to this function as well, as this is critical logic